[AutoDiff] Revamp and fix variedness propagation in activity analysis.

dan-zheng opened the pull request:
Varied values are those that depend on (specific) independent variable,
i.e. function arguments.

For addresses: all projections of a varied address should be varied.
This has special support:
`DifferentiableActivityInfo::propagateVariedInwardsThroughProjections`.

Previously:
- Variedness was propagated by iterating through all instructions in dominance
  order. This is not efficient because irrelevant instructions may be visited.
- For varied addresses, `propagateVariedInwardsThroughProjections` propagated
  variedness one step to projections, but not recursively to users of the
  projections. This caused some values to incorrectly be marked as non-varied.

Now:
- Variedness is propagated by following use-def chains, starting from
  independent variables (function arguments). This is handled by the
  following helpers:
  - `setVariedAndPropagateToUsers(SILValue, unsigned)`: marks a value as varied
    and propagates variedness to users.
  - `propagateVaried(SILInstruction *inst, unsigned)`: propagates
    variedness for the given instruction from operands to results.
- `DifferentiableActivityInfo::propagateVariedInwardsThroughProjections`
  now propagates variedness recursively through projections, which is desired.

Resolves various control flow differentiation correctness issues:
TF-681, TF-781, TF-933.

All of these issues involve incorrect variedness propagation for
initially non-varied addresses.

TF-947 tracks further improvements to varied propagation.
TF-949 tracks similar revamping for usefulness propagation.

---

Visual demonstration using activity info output from `-Xllvm -debug-only=differentiation`:

```swift
// TF-781: check activity analysis for active local address + nested conditionals.
// If an address is varied, all its projections and their users should be varied.

@differentiable(wrt: x)
func TF_781_function(_ x: Float, _ y: Float) -> Float {
  var result = y
  if true {
    if true {
      result = result * x
    }
  }
  return result
}
```

Before:
```
// FIXME(TF-781): `%4 = alloc_stack $Float` is active, so all `begin_access`
// users (and the results of their users, recursively) should also be active.

[AD] Activity info for ${{.*}}TF_781_function{{.*}} at (source=0 parameters=(0))
[ACTIVE] %0 = argument of bb0 : $Float
[USEFUL] %1 = argument of bb0 : $Float
[ACTIVE]   %4 = alloc_stack $Float, var, name "result"
[USEFUL]   %19 = begin_access [read] [static] %4 : $*Float
[USEFUL]   %20 = load [trivial] %19 : $*Float
[ACTIVE]   %23 = apply %22(%20, %0, %18) : $@convention(method) (Float, Float, @thin Float.Type) -> Float
[ACTIVE]   %24 = begin_access [modify] [static] %4 : $*Float
[USEFUL]   %31 = begin_access [read] [static] %4 : $*Float
[USEFUL]   %32 = load [trivial] %31 : $*Float
```

After:
```
[AD] Activity info for ${{.*}}TF_781_function{{.*}} at (source=0 parameters=(0))
[ACTIVE] %0 = argument of bb0 : $Float
[USEFUL] %1 = argument of bb0 : $Float
[ACTIVE]   %4 = alloc_stack $Float, var, name "result"
[ACTIVE]   %19 = begin_access [read] [static] %4 : $*Float
[ACTIVE]   %20 = load [trivial] %19 : $*Float
[ACTIVE]   %23 = apply %22(%20, %0, %18) : $@convention(method) (Float, Float, @thin Float.Type) -> Float
[ACTIVE]   %24 = begin_access [modify] [static] %4 : $*Float
[ACTIVE]   %31 = begin_access [read] [static] %4 : $*Float
[ACTIVE]   %32 = load [trivial] %31 : $*Float
```

dan-zheng committed:
[AutoDiff] Revamp and fix variedness propagation in activity analysis.

Varied values are those that depend on (specific) independent variables,
i.e. function arguments.

For addresses: all projections of a varied address should be varied.
This has special support:
`DifferentiableActivityInfo::propagateVariedInwardsThroughProjections`.

Previously:
- Variedness was propagated by iterating through all instructions in dominance
  order. This is not efficient because irrelevant instructions may be visited.
- For varied addresses, `propagateVariedInwardsThroughProjections` propagated
  variedness one step to projections, but not recursively to users of the
  projections. This caused some values to incorrectly be marked as non-varied.

Now:
- Variedness is propagated by following use-def chains, starting from
  independent variables (function arguments). This is handled by the
  following helpers:
  - `setVariedAndPropagateToUsers(SILValue, unsigned)`: marks a value as varied
    and propagates variedness to users.
  - `propagateVaried(SILInstruction *inst, unsigned)`: propagates
    variedness for the given instruction from operands to results.
- `DifferentiableActivityInfo::propagateVariedInwardsThroughProjections`
  now propagates variedness recursively through projections, which is desired.

Resolves various control flow differentiation correctness issues:
TF-681, TF-781, TF-933.

All of these issues involve incorrect variedness propagation for
initially non-varied addresses.

TF-947 tracks further improvements to varied propagation.
TF-949 tracks similar revamping for usefulness propagation.

dan-zheng commented:
@swift-ci Please test tensorflow

marcrasi commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -2151,6 +2214,25 @@ void DifferentiableActivityInfo::propagateVariedInwardsThroughProjections(
   auto *inst = value->getDefiningInstruction();
   if (!inst || isa<ApplyInst>(inst))
     return;
+  // Recursively propagate varied through users that are projections or
```
Shouldn't the normal `propagateVaried` flow handle everything that this new block of code does, because this is just propagation from operands to results?

marcrasi replied:
Oh, perhaps this is currently necessary because you're calling `setVaried()` a few lines above instead of `setVariedAndPropagateToUsers()`? If changing it to `setVariedAndPropagateToUsers()` is correct, then that seems like a cleaner simpler solution.

dan-zheng replied:
Good suggestion! I made changes in https://github.com/apple/swift/pull/28155/commits/2a0e377133dff15209fe9dbc4923366051a52bc6:
* `propagateVariedInwardsThroughProjections` now calls `setVariedAndPropagateToUsers`, simplifying code.
* Special address-projection logic was removed. It seems no longer necessary.

It's still necessary to keep `propagateVariedInwardsThroughProjections` because it needs to recurse on instruction operands.

rxwei commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -2140,6 +2199,10 @@ void DifferentiableActivityInfo::setUseful(SILValue value,
+  // Check whether value is already varied to prevent infinite recursion.
+  if (isVaried(value, independentVariableIndex))
+    return;
+  // Skip `@noDerivative` struct projections.
```
`s.x` shouldn't be marked active.

dan-zheng replied:
> Should `s.x` be marked active? Does `s.x` get marked active?

`s.x` shouldn't be marked active, please see https://github.com/apple/swift/pull/28155#discussion_r344579232 for context.
For reference, https://github.com/apple/swift/pull/27472 is the patch that changed `s.x` to not be varied. [See test here.](https://github.com/apple/swift/pull/27472/files#diff-da8b8c9bee81d7af57b10d63a2641b4bR47)

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1433,14 +1438,24 @@ class DifferentiableActivityInfo {
+  /// Propagates variedness for the given instruction, from operands to results.
+  void propagateVaried(SILInstruction *inst, unsigned independentVariableIndex);
+  /// Marks the given value as varied and recursively propagates variedness
+  /// inwards (to operands) through projections. Skips `@noDerivative` struct
```
> [...] And what are non-differentiable address projections?

Non-differentiable address projections include:
1. `struct_element_addr` to a `@noDerivative` field.
2. `tuple_element_addr` to an element with a non-differentiable type, like `Int`.

Example of case 2:
```swift
@_silgen_name("nondifferentiable_tuple_element_addr")
@differentiable
func bar(_ x: Float) -> Float {
  var tuple: (Int, Int, Float, Float) = (1, 1, x, x)
  tuple.0 = 1 // non-differentiable `tuple_element_addr`
  tuple.0 = 1 // non-differentiable `tuple_element_addr`
  tuple.3 = x
  return tuple.3
}
```

Such `tuple_element_addr` projections are currently not marked as active (not marked as varied, to be precise):
```
[AD] Activity info for nondifferentiable_tuple_element_addr at (source=0 parameters=(0))
bb0:
[ACTIVE] %0 = argument of bb0 : $Float                     // users: %37, %18, %17, %1
[ACTIVE]   %2 = alloc_stack $(Int, Int, Float, Float), var, name "tuple" // users: %43, %3, %4, %5, %6, %23, %31, %35, %39
[USEFUL]   %3 = tuple_element_addr %2 : $*(Int, Int, Float, Float), 0 // user: %11
[USEFUL]   %4 = tuple_element_addr %2 : $*(Int, Int, Float, Float), 1 // user: %16
[ACTIVE]   %5 = tuple_element_addr %2 : $*(Int, Int, Float, Float), 2 // user: %17
[ACTIVE]   %6 = tuple_element_addr %2 : $*(Int, Int, Float, Float), 3 // user: %18
...
```

> What's the justification for skipping `@noDerivative` struct field projections and non-differentiable address projections? [...] Aren't we supposed to produce a diagnostic when variedness flows through something non-differentiable?

Good question. Here's how the differentiation transform currently works:
* Activity analysis runs before derivative function generation, determining which values in the original function need a derivative.
* Then, [`LinearMapInfo::shouldDifferentiateInstruction`](https://github.com/apple/swift/blob/56548e0ff250ac6b69de5fed023ddf672f761e3a/lib/SILOptimizer/Mandatory/Differentiation.cpp#L1594) uses activity info to determine which instructions to visit during [differential](https://github.com/apple/swift/blob/56548e0ff250ac6b69de5fed023ddf672f761e3a/lib/SILOptimizer/Mandatory/Differentiation.cpp#L4530)/[pullback](https://github.com/apple/swift/blob/56548e0ff250ac6b69de5fed023ddf672f761e3a/lib/SILOptimizer/Mandatory/Differentiation.cpp#L6625) generation.
* If `LinearMapInfo::shouldDifferentiateInstruction` returns true for an instruction, then the instruction is visited during differential/pullback generation. Non-differentiability diagnostics are produced in instruction visitors, e.g. [active `apply` `inout` arguments](https://github.com/apple/swift/blob/56548e0ff250ac6b69de5fed023ddf672f761e3a/lib/SILOptimizer/Mandatory/Differentiation.cpp#L5421). (Non-differentiability diagnostics are also produced in places other than instruction visitors.)

Rather than "producing a diagnostic when variedness flows through something non-differentiable", we currently "avoid marking non-differentiable projections as active, so they're not even visited by differential/pullback generation".

Semantically, this makes sense because non-differentiable projections indeed do not need a derivative, so diagnostics should not be emitted for [writes to non-differentiable projections](https://github.com/apple/swift/pull/27472/files#diff-da8b8c9bee81d7af57b10d63a2641b4bR47).

"Not marking non-differentiable projections as active" also avoids extra downstream work for the compiler and enables nice simplifications, like [making `PullbackEmitter::getAdjointBuffer` no longer need to check for `@noDerivative` or emit diagnostics](https://github.com/apple/swift/pull/27472).

---

I think not marking non-differentiable projections as active is the right direction, and changes like https://github.com/apple/swift/pull/27472 move us towards the desired non-differentiability diagnostics. But the implementation has room for improvement! TF-947 tracks improvements for variedness propagation, additional ideas are welcome!

rxwei replied:
Right, non-differentiable projects (esp. the explicit ones marked by the library with `@noDerivative`) should never be active.

dan-zheng replied:
@marcrasi pointed out that "skipping `@noDerivative` struct projections during variedness propagation" may not be consistent with accessors, which are used in cross-module cases.

I added an item to TF-947 to track checking whether activity analysis is consistent for `struct_extract`/`struct_element_addr` projections and accessor functions to `@noDerivative` properties.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1433,14 +1438,24 @@ class DifferentiableActivityInfo {
   void setVaried(SILValue value, unsigned independentVariableIndex);
```
There's one user of `setVaried` outside of `setVariedAndPropagateToUsers `: `propagateVariedInwardsThroughProjections`. Let's keep `setVaried` for now.

marcrasi replied:
`propagateVariedInwardsThroughProjections` has changed to not use `setVaried` any more, right? So you can remove `setVaried` now.

dan-zheng replied:
> `propagateVariedInwardsThroughProjections` has changed to not use `setVaried` any more, right? So you can remove `setVaried` now.

That's true.

`setVariedAcrossArrayInitialization` also only has one user now, but its body is ad-hoc, so I feel code clarity decreases by inlining it.

How about inlining `setVaried` but not `setVariedAcrossArrayInitialization`?

dan-zheng replied:
Actually, `setVariedAcrossArrayInitialization` seems unnecessary, removing the special support does not affect activity analysis.

Removed both `setVaried` and `setVariedAcrossArrayInitialization` in https://github.com/apple/swift/pull/28191.

dan-zheng committed:
Simplify variedness propagation.

- Change `propagateVaried` to take an `Operand` instead of a `SILInstruction`.
  This makes propagation more precise for `apply`, `br`, and `cond_br` users.
- Remove `propagatedVariedInstructionSets` and related utilities.
  - Instead of keeping track of instructions that have propagated variedness,
    `setVariedAndPropagateToUsers` now skips already-varied values. This is
    simpler and achieves the same effect.
- Simplify `propagateVariedInwardsThroughProjections`.
  - Remove code for skipping non-differentiable address projections.
  - Call `setVariedAndPropagateToUsers` to reuse code.

Addresses feedback from @marcrasi.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1872,6 +1889,131 @@ SILFunction &DifferentiableActivityInfo::getFunction() {
+      if (isVaried(arg, i)) {
+        setDidPropagateVaried(ai, i);
+        for (auto indRes : ai->getIndirectSILResults())
+          setVariedAndPropagateToUsers(indRes, i);
```
Great catch! I'll preemptively change indirect result propagation to use `propagateVariedInwardsThroughProjections` instead of `setVariedAndPropagateToUsers`, for good measure.

---

In practice, I couldn't think of any Swift code where `apply` has an indirect result that's not a local `alloc_stack`. Here's one I tried:

```swift
struct Tuple: Differentiable {
  var x, y: Float
}

@differentiable
func id<T: Differentiable>(_ x: T) -> T { x }

func foo(_ x: Float) -> Float {
  var tuple = Tuple(x: x, y: x)
  return id(tuple.x) // apply with indirect results
}
print(gradient(at: 3, in: foo))
```

```
[AD] Activity info for $foo at (source=0 parameters=(0))
bb0:
[ACTIVE] %0 = argument of bb0 : $Float                     // users: %5, %5, %1
[ACTIVE]   %2 = alloc_stack $Tuple, var, name "tuple"      // users: %19, %6, %8
[USEFUL]   %3 = metatype $@thin Tuple.Type                 // user: %5
[NONE]   // function_ref Tuple.init(x:y:)
  %4 = function_ref @$s6apply25TupleV1x1yACSf_SftcfC : $@convention(method) (Float, Float, @thin Tuple.Type) -> Tuple // user: %5
[ACTIVE]   %5 = apply %4(%0, %0, %3) : $@convention(method) (Float, Float, @thin Tuple.Type) -> Tuple // user: %6
[ACTIVE]   %7 = alloc_stack $Float                         // users: %18, %17, %15
[ACTIVE]   %8 = begin_access [read] [static] %2 : $*Tuple  // users: %11, %9
[ACTIVE]   %9 = struct_element_addr %8 : $*Tuple, #Tuple.x // user: %10
[ACTIVE]   %10 = load [trivial] %9 : $*Float               // user: %13
[ACTIVE]   %12 = alloc_stack $Float                        // users: %16, %15, %13
[NONE]   // function_ref id<A>(_:)
  %14 = function_ref @$s6apply22idyxxs14DifferentiableRzlF : $@convention(thin) <τ_0_0 where τ_0_0 : Differentiable> (@in_guaranteed τ_0_0) -> @out τ_0_0 // user: %15
[NONE]   %15 = apply %14<Float>(%7, %12) : $@convention(thin) <τ_0_0 where τ_0_0 : Differentiable> (@in_guaranteed τ_0_0) -> @out τ_0_0
[ACTIVE]   %17 = load [trivial] %7 : $*Float               // user: %20
```

`%12 = alloc_stack` is the indirect result of the `apply` instruction.

If anyone has ideas for Swift code where SILGen generates `apply` with an indirect result that’s an address projection, please let me know!

dan-zheng replied:
TF-947 tracks checking whether `apply` instruction can have an indirect result that's an address projection.

dan-zheng committed:
Add tuple activity analysis test.

rxwei commented:
Sorry I haven't got a chance to look at this patch.  I should be able to review tomorrow evening.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1872,6 +1889,131 @@ SILFunction &DifferentiableActivityInfo::getFunction() {
+  PROPAGATE_VARIED_THROUGH_STORE(CopyAddr)
+  PROPAGATE_VARIED_THROUGH_STORE(UnconditionalCheckedCastAddr)
+#undef PROPAGATE_VARIED_THROUGH_STORE
+  // Handle `tuple_element_addr`.
```
Added tuple-related activity analysis test in https://github.com/apple/swift/pull/28155/commits/12d908f4d902191bc99365647cd32812c63e460b.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1872,6 +1889,131 @@ SILFunction &DifferentiableActivityInfo::getFunction() {
+    propagateVaried(use->getUser(), independentVariableIndex);
+}
+
+void DifferentiableActivityInfo::propagateVaried(
```
Thanks! Changed `propagateVaried` to take an `Operand *` in https://github.com/apple/swift/pull/28155/commits/2a0e377133dff15209fe9dbc4923366051a52bc6. This makes propagation more precise for `apply`, `br`, and `cond_br` users.

dan-zheng commented on lib/SILOptimizer/Mandatory/Differentiation.cpp:
```diff
@@ -1406,12 +1406,17 @@ class DifferentiableActivityInfo {
+
+  /// The set of instructions that have propagated variedness, indexed by the
+  /// corresponding independent value (input) index.
+  SmallVector<SmallDenseSet<SILInstruction *>, 4>
```
> This data is used to prevent the algorithm from revisiting instructions that it has already marked as varied. Can you achieve the same thing in a simpler way, without storing this data at all, by making `setVariedAndPropagateToUsers` bail out early when the value is already set as varied?

Your first suggestion worked! Done in https://github.com/apple/swift/pull/28155/commits/2a0e377133dff15209fe9dbc4923366051a52bc6, from commit description:

```
    - Remove `propagatedVariedInstructionSets` and related utilities.
      - Instead of keeping track of instructions that have propagated variedness,
        `setVariedAndPropagateToUsers` now skips already-varied values. This is
        simpler and achieves the same effect.
```

dan-zheng committed:
Minor fix.

dan-zheng commented:
I believe I addressed most of @marcrasi's great suggestions.
To unblock progress, I'll trigger tests now and merge if things pass. Very happy to continue discussion and investigate further improvements later!

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow

dan-zheng commented:
@swift-ci Please test tensorflow linux GPU
